torch/_custom_op/autograd.py

```diff
@@ -0,0 +1,222 @@
+import torch
+import torch.utils._pytree as pytree
+from collections import namedtuple
+
+
+# NOTE [CustomOp autograd kernel indirection]
+def autograd_kernel_indirection(custom_op):
+    # TODO(#101191): Use the actual C++ autograd not implemented fallback,
+    # or change the default autograd fallback to the autograd not implemented fallback.
+    def autograd_not_implemented(*args, **kwargs) -> None:
+        if pytree.tree_any(
+            lambda x: isinstance(x, torch.Tensor) and x.requires_grad, (args, kwargs)
+        ):
+            raise RuntimeError("Autograd has not been implemented for operator")
+        guard = torch._C._AutoDispatchBelowAutograd()
+        try:
+            return custom_op(*args, **kwargs)
+        finally:
+            del guard
+
+    def inner(*args, **kwargs):
+        if custom_op._has_impl('autograd'):
+            kernel = custom_op._get_impl('autograd').func
+            return kernel(*args, **kwargs)
+        if custom_op._has_impl('save_for_backward') or custom_op._has_impl('backward'):
```

Comment:
That is the case, yes. I'll add a comment

Reply:
> Is it so that we can avoid passing save_for_backward through the decorator?

Yes. It felt strange that we were passing a function through a decorator, so I split it into two decorators.

> The argument for combining it is that since we always need both, might as well have them registered in the same place.

Yeah I agree that this is an argument for the opposite.